From a43f215efe87a4a518268be7a414c09481005a71 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Thu, 30 Jun 2022 03:17:32 +0000
Subject: [PATCH] xen/arm: Apply MPAM config to dom0less domUs

In dom0less systems, it is not usually very easy to use Xentool to
apply the MPAM configuration to domains. Therefore, the MPAM config
should be applied to the domain at the boot time statically via
device tree properties.

This commit introduces device tree property "xen,domain-mpam-l3-cpbm"
to specify the MPAM cache portion bitmap of the domain L3 cache. The
value will be applied to corresponding MPAM registers at the boot time.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-2794
Change-Id: I8b279e6605fcbf1d723e121fbdf7b42d06eae025
---
 docs/misc/arm/device-tree/booting.txt | 9 +++++++++
 xen/arch/arm/domain_build.c           | 8 ++++++++
 2 files changed, 17 insertions(+)

diff --git a/docs/misc/arm/device-tree/booting.txt b/docs/misc/arm/device-tree/booting.txt
index 87eaa3e254..11e8678115 100644
--- a/docs/misc/arm/device-tree/booting.txt
+++ b/docs/misc/arm/device-tree/booting.txt
@@ -223,6 +223,15 @@ with the following properties:
     the default size of domain P2M pool, i.e. 1MB per guest vCPU plus 4KB
     per MB of guest RAM plus 512KB for guest extended regions.
 
+- xen,domain-mpam-l3-cpbm
+
+    Optional. A 32-bit integer specifying the MPAM cache portion
+    bitmap (CPBM) for the domain L3 cache. Note that MPAM has architecture
+    maximum 32-bit CPBM, but hardware may have different implementations.
+    If the CPBM value given in the device tree exceeds the maximum value
+    of the system, it will be clamped to the maximum value supported
+    by the system.
+
 Under the "xen,domain" compatible node, one or more sub-nodes are present
 for the DomU kernel and ramdisk.
 
diff --git a/xen/arch/arm/domain_build.c b/xen/arch/arm/domain_build.c
index 16435d30b5..93eea0c796 100644
--- a/xen/arch/arm/domain_build.c
+++ b/xen/arch/arm/domain_build.c
@@ -3796,6 +3796,14 @@ static int __init construct_domU(struct domain *d,
     else if ( rc == 0 && !strcmp(dom0less_enhanced, "no-xenstore") )
         kinfo.dom0less_feature = DOM0LESS_ENHANCED_NO_XS;
 
+#ifdef CONFIG_MPAM
+    if ( mpam_enabled )
+    {
+        dt_property_read_u32(node, "xen,domain-mpam-l3-cpbm",
+                             &d->arch.mpam.cfg.l3_cpbm);
+    }
+#endif
+
     if ( vcpu_create(d, 0) == NULL )
         return -ENOMEM;
 
-- 
2.25.1

